research/clustering/information_abstraction.py

```diff
@@ -0,0 +1,439 @@
+        """
+        return np.asarray(list(combinations(self._evals, num_cards)))
+
+
+class GameUtility:
+    """
+    This class takes care of some game related functions
+    """
+    def __init__(self, our_hand: List[int], board: List[int], cards: List[int]):
+
+        self._evaluator = Evaluator()
+        # TODO: this is what takes forever, find a better way
+        self.available_cards = [x for x in cards if x not in board + our_hand]
+        self.our_hand = our_hand
+        self.board = board
+
+    def evaluate_hand(self, hand: List[int]) -> int:
+        """
+        takes a hand
+        :param hand: list of two integers Card.eval_card
+        :return: evaluation of hand
+        """
+        return self._evaluator.evaluate(board=self.board, cards=hand)
+
+    def get_winner(self) -> int:
+        """
+
+        :return: int of win (0), lose (1) or tie (2) - this is an index in the expected hand strength array
+        """
+        our_hand_rank = self.evaluate_hand(self.our_hand)
+        opp_hand_rank = self.evaluate_hand(self.opp_hand)
+        if our_hand_rank > opp_hand_rank:  # maybe some mod magic here
+            return 0
+        elif our_hand_rank < opp_hand_rank:
+            return 1
+        elif our_hand_rank == opp_hand_rank:
+            return 2
+
+    @property
+    def opp_hand(self) -> List[int]:
+        """
+
+        :return: two cards for the opponent (Card.eval_card)
+        """
+        return random.sample(self.available_cards, 2)
+
+
+class InfoSets(ShortDeck):
+    """
+    This class stores combinations of cards (histories) per street (for flop, turn, river)
+    # TODO: should this be isomorphic/lossless to reduce the program run time?
+    """
+    def __init__(self, num_cards):
+        super().__init__(num_cards)
+
+        self.starting_hands = self.get_card_combos(2)
+        self.flop = self.create_info_combos(
+            self.starting_hands, self.get_card_combos(3)
+        )
+        self.turn = self.create_info_combos(self.flop, self.get_card_combos(1))
+        self.river = self.create_info_combos(self.turn, self.get_card_combos(1))
+
+    @staticmethod
+    def create_info_combos(start_combos: np.array, publics: np.array) -> np.ndarray:
+        """
+        Combinations of private information (hole cards) and public information (board)
+        Uses the logic that a AsKsJs on flop with a 10s on turn is different than AsKs10s on flop and Js on turn
+        That logic is used within the literature
+
+        :param start_combos: starting combination of cards (beginning with hole cards)
+        :param publics: np.array of public combinations being added
+        :return: Combinations of private information (hole cards) and public information (board)
+        """
+        our_cards = []
+        for combos in start_combos:
+            for public_combo in publics:
+                # TODO: need a way to create these combos with better performance?
```

Comment:
Maybe there is a more performant way of doing this but I'm impressed - it looks great and it looks like it works following it through with the debugger!

Reply:
ok, yes it's probably fine